src/tools/pkcs11-tool.c

```diff
@@ -1602,22 +1627,133 @@ static int unlock_pin(CK_SLOT_ID slot, CK_SESSION_HANDLE sess, int login_type)
 	return 0;
 }
 
+/* return digest length in bytes */
+static unsigned long figure_pss_salt_length(const int hash) {
+	unsigned long sLen = 0;
+	switch (hash) {
+	case  CKM_SHA_1:
+		sLen = 20;
+		break;
+	case  CKM_SHA256:
+		sLen = 32;
+		break;
+	case  CKM_SHA384:
+		sLen = 48;
+		break;
+	case  CKM_SHA512:
+		sLen = 64;
+		break;
+	default:
+		sLen = 0;
+		break;
```

Comment:
I am not sure if we should silently fall back to `0` length salt in case we got something unknown. It can bite us later (`SHA224`, `SHA3` ?). We should probably exit here.

Reply:
I see your point. But I'm not sure how we'd do a clean exit here without overly complicating things on the caller side.

As for SHA-3 and SHA224 - why don't we add them right now and be done with it?

Reply:
All the way around `pkcs11-tool`, the we call `util_fatal()`, which calls exit. It is not nice, but it does its job for these unexpected situations. I 

We can add `CKM_SHA224` straight away, but there is no `CKM_` for SHA3 yet in latest PKCS#11 standard. And when it will be there, we will forget about this switch.

I will add a commits addressing this.

Reply:
OK, thanks. Makes sense to me. For an executable like `pkcs11-tool` using `util_fatal()` should be OK.

Reply:
Darn... We must be using different PKCS#11 include files?!
```
. . . . .
  CC       cardos-tool.o
pkcs11-tool.c:1637:8: error: use of undeclared identifier 'CKM_SHA224'
        case  CKM_SHA224:
              ^
1 error generated.
make[3]: *** [pkcs11-tool.o] Error 1
. . . . .
```

Reply:
Nope. I just did not check it compiles. I checked only the PKCS#11 specification, which defines them:

http://docs.oasis-open.org/pkcs11/pkcs11-curr/v2.40/cs01/pkcs11-curr-v2.40-cs01.html#_Toc399398977

I added them in b051d3b0 from the above source so it should build fine.

Reply:
Yep, thanks. Now everything compiles OK. ;-)